File: ru/codes/python/chapter_tree/binary_tree_bfs.py
"""
File: binary_tree_bfs.py
Created Time: 2022-12-20
Author: a16su (lpluls001@gmail.com)
"""

import sys
from pathlib import Path

sys.path.append(str(Path(__file__).parent.parent))
from modules import TreeNode, list_to_tree, print_tree
from collections import deque


def level_order(root: TreeNode | None) -> list[int]:
    """Обход в ширину"""
    # Инициализировать очередь и добавить корневой узел
    queue: deque[TreeNode] = deque()
    queue.append(root)
    # Инициализировать список для хранения последовательности обхода
    res = []
    while queue:
        node: TreeNode = queue.popleft()  # Извлечение из очереди
        res.append(node.val)  # Сохранить значение узла
        if node.left is not None:
            queue.append(node.left)  # Поместить левый дочерний узел в очередь
        if node.right is not None:
            queue.append(node.right)  # Поместить правый дочерний узел в очередь
    return res


"""Driver Code"""
if __name__ == "__main__":
    # Инициализировать двоичное дерево
    # Здесь используется функция, напрямую строящая двоичное дерево из массива
    root: TreeNode = list_to_tree(arr=[1, 2, 3, 4, 5, 6, 7])
    print("\nИнициализация двоичного дерева\n")
    print_tree(root)

    # Обход в ширину
    res: list[int] = level_order(root)
    print("\nПоследовательность печати узлов при обходе в ширину = ", res)

File: ru/codes/csharp/chapter_tree/binary_tree_bfs.cs
/**
 * File: binary_tree_bfs.cs
 * Created Time: 2022-12-23
 * Author: haptear (haptear@hotmail.com)
 */

namespace hello_algo.chapter_tree;

public class binary_tree_bfs {

    /* Обход в ширину */
    List<int> LevelOrder(TreeNode root) {
        // Инициализировать очередь и добавить корневой узел
        Queue<TreeNode> queue = new();
        queue.Enqueue(root);
        // Инициализировать список для хранения последовательности обхода
        List<int> list = [];
        while (queue.Count != 0) {
            TreeNode node = queue.Dequeue(); // Извлечение из очереди
            list.Add(node.val!.Value);       // Сохранить значение узла
            if (node.left != null)
                queue.Enqueue(node.left);    // Поместить левый дочерний узел в очередь
            if (node.right != null)
                queue.Enqueue(node.right);   // Поместить правый дочерний узел в очередь
        }
        return list;
    }

    [Test]
    public void Test() {
        /* Инициализация двоичного дерева */
        // Здесь используется функция, напрямую строящая двоичное дерево из массива
        TreeNode? root = TreeNode.ListToTree([1, 2, 3, 4, 5, 6, 7]);
        Console.WriteLine("\nИнициализация двоичного дерева\n");
        PrintUtil.PrintTree(root);

        List<int> list = LevelOrder(root!);
        Console.WriteLine("\nПоследовательность печати узлов при обходе в ширину = " + string.Join(",", list));
    }
}
